contracts/random/RandomValues.sol

```diff
@@ -0,0 +1,123 @@
+    mapping(uint256 requestId => address randomSource) private randCreationRequestsSource;
+
+    // Each request has a unique request id. The id is the current value
+    // of nextNonce. nextNonce is incremented for each request.
+    uint256 private nextNonce;
+
+    /**
+     * @notice Set the address of the random seed provider.
+     * @param _randomSeedProvider Address of random seed provider.
+     */
+    constructor(address _randomSeedProvider) {
+        randomSeedProvider = RandomSeedProvider(_randomSeedProvider);
+    }
+
+    /**
+     * @notice Register a request to generate a random value. This function should be called
+     *      when a game player has purchased an item the value of which is based on a random value.
+     * @return _randomRequestId A value that needs to be presented when fetching the random
+     *      value with fetchRandom.
+     */
+    function _requestRandomValueCreation() internal returns (uint256 _randomRequestId) {
+        uint256 randomFulfillmentIndex;
+        address randomSource;
+        (randomFulfillmentIndex, randomSource) = randomSeedProvider.requestRandomSeed();
+        _randomRequestId = nextNonce++;
+        randCreationRequests[_randomRequestId] = randomFulfillmentIndex;
+        randCreationRequestsSource[_randomRequestId] = randomSource;
+    }
+
+    /**
+     * @notice Fetch a random value that was requested using _requestRandomValueCreation.
+     * @dev The value is customised to this game, the game player, and the request by the game player.
+     *      This level of personalisation ensures that no two players end up with the same random value
+     *      and no game player will have the same random value twice.
+     * @param _randomRequestId The value returned by _requestRandomValueCreation.
+     * @return _randomValue The random number that the game can use.
+     */
+    function _fetchRandom(uint256 _randomRequestId) internal returns (bytes32 _randomValue) {
+        return _fetchPersonalisedSeed(_randomRequestId);
+    }
+
+    /**
+     * @notice Fetch a set of random values that were requested using _requestRandomValueCreation.
+     * @dev The values are customised to this game, the game player, and the request by the game player.
+     *      This level of personalisation ensures that no two players end up with the same random value
+     *      and no game player will have the same random value twice.
+     * @param _randomRequestId The value returned by _requestRandomValueCreation.
+     * @param _size The size of the array to return.
+     * @return _randomValues An array of random values derived from a single random value.
+     */
+    function _fetchRandomValues(
+        uint256 _randomRequestId,
+        uint256 _size
+    ) internal returns (bytes32[] memory _randomValues) {
+        bytes32 seed = _fetchPersonalisedSeed(_randomRequestId);
+
+        _randomValues = new bytes32[](_size);
+        for (uint256 i = 0; i < _size; i++) {
```

Comment:
One thing to note: if a player gets a favourable value, they may attempt multiple txs in the same block. We should consider limiting a fetched value to oncer per block

Reply:
Alternatively, make it clear to developer that they need to manage that scenario in their Game contracts

Reply:
That scenario is catered for. The game will call _requestRandomValueCreation twice in the same block. They will get back two different _randomRequestId values. When they later call _fetchRandom or _fetchRandomValues, they will use the same seed from the RandomSeedProvider, but will then personalise it based on the random request id (and the game contract and the msg.sender).